File: src/test/java/no/nav/sosialhjelp/soknad/business/batch/oppgave/fiks/FiksSenderTest.java
package no.nav.sosialhjelp.soknad.business.batch.oppgave.fiks;

import no.ks.fiks.svarut.klient.model.Digitaladresse;
import no.ks.fiks.svarut.klient.model.Dokument;
import no.ks.fiks.svarut.klient.model.Forsendelse;
import no.nav.sbl.soknadsosialhjelp.soknad.JsonData;
import no.nav.sbl.soknadsosialhjelp.soknad.JsonDriftsinformasjon;
import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad;
import no.nav.sbl.soknadsosialhjelp.soknad.JsonSoknad;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomi;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomiopplysninger;
import no.nav.sbl.soknadsosialhjelp.vedlegg.JsonFiler;
import no.nav.sbl.soknadsosialhjelp.vedlegg.JsonVedlegg;
import no.nav.sbl.soknadsosialhjelp.vedlegg.JsonVedleggSpesifikasjon;
import no.nav.sosialhjelp.soknad.business.InnsendingService;
import no.nav.sosialhjelp.soknad.business.pdfmedpdfbox.SosialhjelpPdfGenerator;
import no.nav.sosialhjelp.soknad.client.svarut.SvarUtService;
import no.nav.sosialhjelp.soknad.consumer.fiks.DokumentKrypterer;
import no.nav.sosialhjelp.soknad.domain.OpplastetVedlegg;
import no.nav.sosialhjelp.soknad.domain.SendtSoknad;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid;
import no.nav.sosialhjelp.soknad.domain.VedleggType;
import no.nav.sosialhjelp.soknad.domain.Vedleggstatus;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoSettings;
import org.mockito.quality.Strictness;

import java.io.InputStream;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.UUID;

import static no.nav.sosialhjelp.soknad.business.batch.oppgave.fiks.FiksSender.ETTERSENDELSE_TIL_NAV;
import static no.nav.sosialhjelp.soknad.business.batch.oppgave.fiks.FiksSender.SOKNAD_TIL_NAV;
import static no.nav.sosialhjelp.soknad.business.service.soknadservice.SoknadService.createEmptyJsonInternalSoknad;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.anyBoolean;
import static org.mockito.Mockito.anyString;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@MockitoSettings(strictness = Strictness.LENIENT)
class FiksSenderTest {

    private static final String FIKSFORSENDELSE_ID = UUID.randomUUID().toString();
    private static final String FILNAVN = "filnavn.jpg";
    private static final String ORGNUMMER = "9999";
    private static final String NAVENHETSNAVN = "NAV Sagene";
    private static final String BEHANDLINGSID = "12345";
    private static final String EIER = "12345678910";
    @Mock
    private DokumentKrypterer dokumentKrypterer;
    @Mock
    private InnsendingService innsendingService;
    @Mock
    private SosialhjelpPdfGenerator sosialhjelpPdfGenerator;
    @Mock
    private SvarUtService svarUtService;

    private FiksSender fiksSender;

    @BeforeEach
    public void setUp() {
        System.clearProperty("environment.name");
        when(dokumentKrypterer.krypterData(any())).thenReturn(new byte[]{3, 2, 1});
        when(innsendingService.finnSendtSoknadForEttersendelse(any(SoknadUnderArbeid.class)))
                .thenReturn(new SendtSoknad().withFiksforsendelseId(FIKSFORSENDELSE_ID));
        when(innsendingService.hentSoknadUnderArbeid(anyString(), anyString())).thenReturn(new SoknadUnderArbeid());
        when(sosialhjelpPdfGenerator.generate(any(JsonInternalSoknad.class), anyBoolean())).thenReturn(new byte[]{1, 2, 3});
        when(sosialhjelpPdfGenerator.generateEttersendelsePdf(any(JsonInternalSoknad.class), anyString())).thenReturn(new byte[]{1, 2, 3});
        when(sosialhjelpPdfGenerator.generateBrukerkvitteringPdf()).thenReturn(new byte[]{1, 2, 3});

        fiksSender = new FiksSender(dokumentKrypterer, innsendingService, sosialhjelpPdfGenerator, true, svarUtService);
    }

    @Test
    void createForsendelseSetterRiktigInfoPaForsendelsenMedKryptering() {
        when(innsendingService.hentSoknadUnderArbeid(anyString(), anyString()))
                .thenReturn(new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER)));
        SendtSoknad sendtSoknad = lagSendtSoknad();

        var filnavnInputStreamMap = new HashMap<String, InputStream>();

        Forsendelse forsendelse = fiksSender.createForsendelse(sendtSoknad, filnavnInputStreamMap);

        Digitaladresse adresse = forsendelse.getMottaker().getDigitalAdresse();
        assertThat(adresse.getOrganisasjonsNummer()).isEqualTo(ORGNUMMER);
        assertThat(forsendelse.getMottaker().getPostAdresse().getNavn()).isEqualTo(NAVENHETSNAVN);
        assertThat(forsendelse.getAvgivendeSystem()).isEqualTo("digisos_avsender");
        assertThat(forsendelse.getForsendelsesType()).isEqualTo("nav.digisos");
        assertThat(forsendelse.getEksternReferanse()).isEqualTo(BEHANDLINGSID);
        assertThat(forsendelse.isKunDigitalLevering()).isFalse();
        assertThat(forsendelse.getUtskriftsKonfigurasjon().isTosidig()).isTrue();
        assertThat(forsendelse.isKryptert()).isTrue();
        assertThat(forsendelse.isKrevNiva4Innlogging()).isTrue();
        assertThat(forsendelse.getSvarPaForsendelse()).isNull();
        assertThat(forsendelse.getDokumenter()).hasSize(5);
        assertThat(forsendelse.getMetadataFraAvleverendeSystem().getDokumentetsDato()).isNotNull();
        verify(dokumentKrypterer, times(5)).krypterData(any());
    }

    @Test
    void opprettForsendelseSetterRiktigInfoPaForsendelsenUtenKryptering() {
        fiksSender = new FiksSender(dokumentKrypterer, innsendingService, sosialhjelpPdfGenerator, false, svarUtService);

        when(innsendingService.hentSoknadUnderArbeid(anyString(), anyString()))
                .thenReturn(new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER)).withEier(EIER));
        SendtSoknad sendtSoknad = lagSendtSoknad();

        var filnavnInputStreamMap = new HashMap<String, InputStream>();

        Forsendelse forsendelse = fiksSender.createForsendelse(sendtSoknad, filnavnInputStreamMap);

        assertThat(forsendelse.isKryptert()).isFalse();
        assertThat(forsendelse.isKrevNiva4Innlogging()).isFalse();
        verify(dokumentKrypterer, never()).krypterData(any());
    }

    @Test
    void createForsendelseSetterRiktigTittelForNySoknad() {
        when(innsendingService.hentSoknadUnderArbeid(anyString(), anyString()))
                .thenReturn(new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER)).withEier(EIER));
        SendtSoknad sendtSoknad = lagSendtSoknad();

        var filnavnInputStreamMap = new HashMap<String, InputStream>();

        Forsendelse forsendelse = fiksSender.createForsendelse(sendtSoknad, filnavnInputStreamMap);

        assertThat(forsendelse.getTittel()).isEqualTo(SOKNAD_TIL_NAV);
    }

    @Test
    void createForsendelseSetterRiktigTittelForEttersendelse() {
        when(innsendingService.hentSoknadUnderArbeid(anyString(), anyString())).thenReturn(new SoknadUnderArbeid()
                .withTilknyttetBehandlingsId("12345")
                .withJsonInternalSoknad(lagInternalSoknadForEttersending())
                .withEier(EIER));
        //when(any(SoknadUnderArbeid.class).getJsonInternalSoknad()).thenReturn(lagInternalSoknadForEttersending());
        SendtSoknad sendtSoknad = lagSendtSoknad().withTilknyttetBehandlingsId("12345");

        var filnavnInputStreamMap = new HashMap<String, InputStream>();

        Forsendelse forsendelse = fiksSender.createForsendelse(sendtSoknad, filnavnInputStreamMap);

        assertThat(forsendelse.getTittel()).isEqualTo(ETTERSENDELSE_TIL_NAV);
    }

    @Test
    void opprettForsendelseForEttersendelseUtenSvarPaForsendelseSkalFeile() {
        when(innsendingService.hentSoknadUnderArbeid(anyString(), anyString()))
                .thenReturn(new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER)).withEier(EIER));
        when(innsendingService.finnSendtSoknadForEttersendelse(any(SoknadUnderArbeid.class))).thenReturn(new SendtSoknad()
                .withFiksforsendelseId(null));
        SendtSoknad sendtEttersendelse = lagSendtEttersendelse();

        var filnavnInputStreamMap = new HashMap<String, InputStream>();

        assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(() -> fiksSender.createForsendelse(sendtEttersendelse, filnavnInputStreamMap));
    }

    @Test
    void hentDokumenterFraSoknadReturnererFireDokumenterForSoknadUtenVedlegg() {
        var filnavnInputStreamMap = new HashMap<String, InputStream>();

        List<Dokument> fiksDokumenter = fiksSender.hentDokumenterFraSoknad(
                new SoknadUnderArbeid()
                        .withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER)),
                filnavnInputStreamMap);

        assertThat(fiksDokumenter).hasSize(5);
        assertThat(fiksDokumenter.get(0).getFilnavn()).isEqualTo("soknad.json");
        assertThat(fiksDokumenter.get(1).getFilnavn()).isEqualTo("Soknad.pdf");
        assertThat(fiksDokumenter.get(2).getFilnavn()).isEqualTo("vedlegg.json");
        assertThat(fiksDokumenter.get(3).getFilnavn()).isEqualTo("Soknad-juridisk.pdf");
        assertThat(fiksDokumenter.get(4).getFilnavn()).isEqualTo("Brukerkvittering.pdf");
    }

    @Test
    void hentDokumenterFraSoknadReturnererTreDokumenterForEttersendingMedEtVedlegg() {
        var filnavnInputStreamMap = new HashMap<String, InputStream>();
        when(innsendingService.hentAlleOpplastedeVedleggForSoknad(any(SoknadUnderArbeid.class))).thenReturn(lagOpplastetVedlegg());

        List<Dokument> fiksDokumenter = fiksSender.hentDokumenterFraSoknad(
                new SoknadUnderArbeid()
                    .withTilknyttetBehandlingsId("123")
                    .withJsonInternalSoknad(lagInternalSoknadForEttersending()),
                filnavnInputStreamMap);

        assertThat(fiksDokumenter).hasSize(4);
        assertThat(fiksDokumenter.get(0).getFilnavn()).isEqualTo("ettersendelse.pdf");
        assertThat(fiksDokumenter.get(1).getFilnavn()).isEqualTo("vedlegg.json");
        assertThat(fiksDokumenter.get(2).getFilnavn()).isEqualTo("Brukerkvittering.pdf");
        assertThat(fiksDokumenter.get(3).getFilnavn()).isEqualTo(FILNAVN);
    }

    @Test
    void hentDokumenterFraSoknadKasterFeilHvisSoknadManglerForNySoknad() {
        var filnavnInputStreamMap = new HashMap<String, InputStream>();
        assertThatExceptionOfType(RuntimeException.class)
                .isThrownBy(() -> fiksSender.hentDokumenterFraSoknad(new SoknadUnderArbeid(), filnavnInputStreamMap));
    }

    @Test
    void hentDokumenterFraSoknadKasterFeilHvisVedleggManglerForEttersending() {
        var filnavnInputStreamMap = new HashMap<String, InputStream>();
        assertThatExceptionOfType(RuntimeException.class)
                .isThrownBy(() -> fiksSender.hentDokumenterFraSoknad(new SoknadUnderArbeid().withTilknyttetBehandlingsId("123"), filnavnInputStreamMap));
    }

    private JsonInternalSoknad lagInternalSoknadForEttersending() {
        List<JsonFiler> jsonFiler = new ArrayList<>();
        jsonFiler.add(new JsonFiler().withFilnavn(FILNAVN).withSha512("sha512"));
        List<JsonVedlegg> jsonVedlegg = new ArrayList<>();
        jsonVedlegg.add(new JsonVedlegg()
                .withStatus(Vedleggstatus.LastetOpp.name())
                .withType("type")
                .withTilleggsinfo("tilleggsinfo")
                .withFiler(jsonFiler));
        return new JsonInternalSoknad()
                .withVedlegg(new JsonVedleggSpesifikasjon().withVedlegg(jsonVedlegg))
                .withSoknad(new JsonSoknad()
                        .withDriftsinformasjon(new JsonDriftsinformasjon())
                        .withData(new JsonData()
                                .withOkonomi(new JsonOkonomi()
                                .withOpplysninger(new JsonOkonomiopplysninger()))));
    }

    private List<OpplastetVedlegg> lagOpplastetVedlegg() {
        List<OpplastetVedlegg> opplastedeVedlegg = new ArrayList<>();
        opplastedeVedlegg.add(new OpplastetVedlegg()
                .withFilnavn(FILNAVN)
                .withSha512("sha512")
                .withVedleggType(new VedleggType("type|tilleggsinfo"))
                .withData(new byte[]{1, 2, 3}));
        return opplastedeVedlegg;
    }

    private SendtSoknad lagSendtSoknad() {
        return new SendtSoknad()
                .withBehandlingsId(BEHANDLINGSID)
                .withOrgnummer(ORGNUMMER)
                .withNavEnhetsnavn(NAVENHETSNAVN)
                .withEier(EIER)
                .withBrukerFerdigDato(LocalDateTime.now());
    }

    private SendtSoknad lagSendtEttersendelse() {
        return lagSendtSoknad().withTilknyttetBehandlingsId("soknadId");
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/innsending/svarut/FiksSenderTest.kt
package no.nav.sosialhjelp.soknad.innsending.svarut

import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkObject
import io.mockk.unmockkObject
import io.mockk.verify
import no.nav.sbl.soknadsosialhjelp.soknad.JsonData
import no.nav.sbl.soknadsosialhjelp.soknad.JsonDriftsinformasjon
import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad
import no.nav.sbl.soknadsosialhjelp.soknad.JsonSoknad
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomi
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomiopplysninger
import no.nav.sbl.soknadsosialhjelp.vedlegg.JsonFiler
import no.nav.sbl.soknadsosialhjelp.vedlegg.JsonVedlegg
import no.nav.sbl.soknadsosialhjelp.vedlegg.JsonVedleggSpesifikasjon
import no.nav.sosialhjelp.soknad.app.MiljoUtils
import no.nav.sosialhjelp.soknad.db.repositories.opplastetvedlegg.OpplastetVedlegg
import no.nav.sosialhjelp.soknad.db.repositories.opplastetvedlegg.OpplastetVedleggType
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.SoknadMetadata
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.Vedleggstatus
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeid
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidStatus
import no.nav.sosialhjelp.soknad.innsending.InnsendingService
import no.nav.sosialhjelp.soknad.innsending.SoknadServiceOld.Companion.createEmptyJsonInternalSoknad
import no.nav.sosialhjelp.soknad.innsending.svarut.client.SvarUtService
import no.nav.sosialhjelp.soknad.pdf.SosialhjelpPdfGenerator
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import java.io.InputStream
import java.time.LocalDateTime
import java.util.UUID

internal class FiksSenderTest {
    private val dokumentKrypterer: DokumentKrypterer = mockk()
    private val innsendingService: InnsendingService = mockk()
    private val sosialhjelpPdfGenerator: SosialhjelpPdfGenerator = mockk()
    private val svarUtService: SvarUtService = mockk()

    private val fiksSender = FiksSender(dokumentKrypterer, innsendingService, sosialhjelpPdfGenerator, true, svarUtService)

    @BeforeEach
    fun setUp() {
        mockkObject(MiljoUtils)
        every { MiljoUtils.isNonProduction() } returns false
        every { dokumentKrypterer.krypterData(any()) } returns byteArrayOf(3, 2, 1)
        val soknadMetadata = lagSoknadMetadata()
        soknadMetadata.fiksForsendelseId = FIKSFORSENDELSE_ID
        every { innsendingService.finnFiksForsendelseIdForEttersendelse(any()) } returns FIKSFORSENDELSE_ID
        every { innsendingService.hentSoknadUnderArbeid(any(), any()) } returns mockk()
        every { sosialhjelpPdfGenerator.generate(any(), any()) } returns byteArrayOf(1, 2, 3)
        every { sosialhjelpPdfGenerator.generateEttersendelsePdf(any(), any()) } returns byteArrayOf(1, 2, 3)
        every { sosialhjelpPdfGenerator.generateBrukerkvitteringPdf() } returns byteArrayOf(1, 2, 3)
        every { innsendingService.hentAlleOpplastedeVedleggForSoknad(any()) } returns emptyList()
    }

    @AfterEach
    internal fun tearDown() {
        unmockkObject(MiljoUtils)
    }

    @Test
    fun createForsendelseSetterRiktigInfoPaForsendelsenMedKryptering() {
        every { innsendingService.hentSoknadUnderArbeid(any(), any()) } returns createSoknadUnderArbeid()

        val soknadMetadata = lagSoknadMetadata()
        val filnavnInputStreamMap = HashMap<String, InputStream>()
        val forsendelse = fiksSender.createForsendelse(soknadMetadata, filnavnInputStreamMap)

        val adresse = forsendelse.mottaker.digitalAdresse
        assertThat(adresse.organisasjonsNummer).isEqualTo(ORGNUMMER)
        assertThat(forsendelse.mottaker.postAdresse.navn).isEqualTo(NAVENHETSNAVN)
        assertThat(forsendelse.avgivendeSystem).isEqualTo("digisos_avsender")
        assertThat(forsendelse.forsendelsesType).isEqualTo("nav.digisos")
        assertThat(forsendelse.eksternReferanse).isEqualTo(BEHANDLINGSID)
        assertThat(forsendelse.isKunDigitalLevering).isFalse
        assertThat(forsendelse.utskriftsKonfigurasjon.isTosidig).isTrue
        assertThat(forsendelse.isKryptert).isTrue
        assertThat(forsendelse.isKrevNiva4Innlogging).isTrue
        assertThat(forsendelse.svarPaForsendelse).isNull()
        assertThat(forsendelse.dokumenter).hasSize(5)
        assertThat(forsendelse.metadataFraAvleverendeSystem.dokumentetsDato).isNotNull
        verify(exactly = 5) { dokumentKrypterer.krypterData(any()) }
    }

    @Test
    fun opprettForsendelseSetterRiktigInfoPaForsendelsenUtenKryptering() {
        every { MiljoUtils.isNonProduction() } returns true
        every { MiljoUtils.environmentName } returns "test"
        val fiksSenderUtenKryptering = FiksSender(dokumentKrypterer, innsendingService, sosialhjelpPdfGenerator, false, svarUtService)
        every { innsendingService.hentSoknadUnderArbeid(any(), any()) } returns createSoknadUnderArbeid()

        val soknadMetadata = lagSoknadMetadata()
        val filnavnInputStreamMap = HashMap<String, InputStream>()
        val forsendelse = fiksSenderUtenKryptering.createForsendelse(soknadMetadata, filnavnInputStreamMap)

        assertThat(forsendelse.isKryptert).isFalse
        assertThat(forsendelse.isKrevNiva4Innlogging).isFalse
        verify(exactly = 0) { dokumentKrypterer.krypterData(any()) }
    }

    @Test
    fun createForsendelseSetterRiktigTittelForNySoknad() {
        every { innsendingService.hentSoknadUnderArbeid(any(), any()) } returns createSoknadUnderArbeid()

        val soknadMetadata = lagSoknadMetadata()
        val filnavnInputStreamMap = HashMap<String, InputStream>()
        val forsendelse = fiksSender.createForsendelse(soknadMetadata, filnavnInputStreamMap)

        assertThat(forsendelse.tittel).isEqualTo(FiksSender.SOKNAD_TIL_NAV)
    }

    @Test
    @Disabled("Gjelder ettersendelse")
    fun createForsendelseSetterRiktigTittelForEttersendelse() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.tilknyttetBehandlingsId = "12345"
        soknadUnderArbeid.jsonInternalSoknad = lagInternalSoknadForEttersending()
        every { innsendingService.hentSoknadUnderArbeid(any(), any()) } returns soknadUnderArbeid

        val soknadMetadata = lagSoknadMetadata()
        soknadMetadata.tilknyttetBehandlingsId = "12345"

        val filnavnInputStreamMap = HashMap<String, InputStream>()
        val forsendelse = fiksSender.createForsendelse(soknadMetadata, filnavnInputStreamMap)

        assertThat(forsendelse.tittel).isEqualTo(FiksSender.ETTERSENDELSE_TIL_NAV)
    }

    @Test
    fun opprettForsendelseForEttersendelseUtenSvarPaForsendelseSkalFeile() {
        every { innsendingService.hentSoknadUnderArbeid(any(), any()) } returns createSoknadUnderArbeid()
        every { innsendingService.finnFiksForsendelseIdForEttersendelse(any()) } returns null

        val soknadMetadataEttersendelse = lagSoknadMetadataEttersendelse()
        val filnavnInputStreamMap = HashMap<String, InputStream>()

        assertThatExceptionOfType(IllegalStateException::class.java)
            .isThrownBy { fiksSender.createForsendelse(soknadMetadataEttersendelse, filnavnInputStreamMap) }
    }

    @Test
    fun hentDokumenterFraSoknadReturnererFireDokumenterForSoknadUtenVedlegg() {
        val filnavnInputStreamMap = HashMap<String, InputStream>()
        val fiksDokumenter = fiksSender.hentDokumenterFraSoknad(createSoknadUnderArbeid(), filnavnInputStreamMap)
        assertThat(fiksDokumenter).hasSize(5)
        assertThat(fiksDokumenter[0].filnavn).isEqualTo("soknad.json")
        assertThat(fiksDokumenter[1].filnavn).isEqualTo("Soknad.pdf")
        assertThat(fiksDokumenter[2].filnavn).isEqualTo("vedlegg.json")
        assertThat(fiksDokumenter[3].filnavn).isEqualTo("Soknad-juridisk.pdf")
        assertThat(fiksDokumenter[4].filnavn).isEqualTo("Brukerkvittering.pdf")
    }

    @Test
    @Disabled("Gjelder ettersendelse")
    fun hentDokumenterFraSoknadReturnererTreDokumenterForEttersendingMedEtVedlegg() {
        val filnavnInputStreamMap = HashMap<String, InputStream>()
        every { innsendingService.hentAlleOpplastedeVedleggForSoknad(any()) } returns lagOpplastetVedlegg()

        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.tilknyttetBehandlingsId = "123"
        soknadUnderArbeid.jsonInternalSoknad = lagInternalSoknadForEttersending()
        val fiksDokumenter = fiksSender.hentDokumenterFraSoknad(soknadUnderArbeid, filnavnInputStreamMap)
        assertThat(fiksDokumenter).hasSize(4)
        assertThat(fiksDokumenter[0].filnavn).isEqualTo("ettersendelse.pdf")
        assertThat(fiksDokumenter[1].filnavn).isEqualTo("vedlegg.json")
        assertThat(fiksDokumenter[2].filnavn).isEqualTo("Brukerkvittering.pdf")
        assertThat(fiksDokumenter[3].filnavn).isEqualTo(FILNAVN)
    }

    @Test
    fun hentDokumenterFraSoknadKasterFeilHvisSoknadManglerForNySoknad() {
        val filnavnInputStreamMap = HashMap<String, InputStream>()
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.jsonInternalSoknad?.soknad = null
        assertThatExceptionOfType(RuntimeException::class.java)
            .isThrownBy { fiksSender.hentDokumenterFraSoknad(soknadUnderArbeid, filnavnInputStreamMap) }
    }

    @Test
    fun hentDokumenterFraSoknadKasterFeilHvisVedleggManglerForEttersending() {
        val filnavnInputStreamMap = HashMap<String, InputStream>()
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.tilknyttetBehandlingsId = "123"
        soknadUnderArbeid.jsonInternalSoknad?.vedlegg = null
        assertThatExceptionOfType(RuntimeException::class.java)
            .isThrownBy { fiksSender.hentDokumenterFraSoknad(soknadUnderArbeid, filnavnInputStreamMap) }
    }

    private fun lagInternalSoknadForEttersending(): JsonInternalSoknad {
        val jsonFiler = mutableListOf<JsonFiler>()
        jsonFiler.add(JsonFiler().withFilnavn(FILNAVN).withSha512("sha512"))
        val jsonVedlegg = mutableListOf<JsonVedlegg>()
        jsonVedlegg.add(
            JsonVedlegg()
                .withStatus(Vedleggstatus.LastetOpp.name)
                .withType("type")
                .withTilleggsinfo("tilleggsinfo")
                .withFiler(jsonFiler),
        )
        return JsonInternalSoknad()
            .withVedlegg(JsonVedleggSpesifikasjon().withVedlegg(jsonVedlegg))
            .withSoknad(
                JsonSoknad()
                    .withDriftsinformasjon(JsonDriftsinformasjon())
                    .withData(
                        JsonData()
                            .withOkonomi(
                                JsonOkonomi()
                                    .withOpplysninger(JsonOkonomiopplysninger()),
                            ),
                    ),
            )
    }

    private fun lagOpplastetVedlegg(): List<OpplastetVedlegg> {
        return mutableListOf(
            OpplastetVedlegg(
                eier = "eier",
                vedleggType = OpplastetVedleggType("type|tilleggsinfo"),
                data = byteArrayOf(1, 2, 3),
                soknadId = 123L,
                filnavn = FILNAVN,
                sha512 = "sha512",
            ),
        )
    }

    private fun lagSoknadMetadata(): SoknadMetadata {
        return SoknadMetadata(
            id = 1L,
            behandlingsId = BEHANDLINGSID,
            tilknyttetBehandlingsId = null,
            fnr = EIER,
            fiksForsendelseId = null,
            orgnr = ORGNUMMER,
            navEnhet = NAVENHETSNAVN,
            opprettetDato = LocalDateTime.now(),
            sistEndretDato = LocalDateTime.now(),
            innsendtDato = null,
        )
    }

    private fun lagSoknadMetadataEttersendelse(): SoknadMetadata {
        val soknadMetadata = lagSoknadMetadata()
        soknadMetadata.tilknyttetBehandlingsId = "soknadId"
        return soknadMetadata
    }

    companion object {
        private val FIKSFORSENDELSE_ID = UUID.randomUUID().toString()
        private const val FILNAVN = "filnavn.jpg"
        private const val ORGNUMMER = "9999"
        private const val NAVENHETSNAVN = "NAV Sagene"
        private const val BEHANDLINGSID = "12345"
        private const val EIER = "12345678910"

        private fun createSoknadUnderArbeid(): SoknadUnderArbeid {
            return SoknadUnderArbeid(
                versjon = 1L,
                behandlingsId = "behandlingsid",
                tilknyttetBehandlingsId = null,
                eier = EIER,
                jsonInternalSoknad = createEmptyJsonInternalSoknad(EIER),
                status = SoknadUnderArbeidStatus.UNDER_ARBEID,
                opprettetDato = LocalDateTime.now(),
                sistEndretDato = LocalDateTime.now(),
            )
        }
    }
}
